timer: fix timeout race condition with smp

chen-png opened the pull request:
before running timer's timeout function, we need to make
sure that those threads waiting on this timer have been
added into the timer's wait queue, so add operations to
use timer lock to mask interrupts in z_timer_expiration_handler
function to synchronize timer's wait queue.

fix: #37304 

Signed-off-by: Chen Peng1 <peng1.chen@intel.com>

enjiamai reviewed (approved):
Looks good. Is it possible split it into 2 commits, if possible?

- Make the spinlock as global and use it. 

- Rename them from "lock" to "timer_lock"

This is just a suggestion. Will make the changes more clear.

chen-png commented:
hi @dcpleung @andyross Could you help take a look of this? thanks.

andyross commented:
Can you explain exactly what the sequencing is here?  Like, "CPU A does this" "CPU B does this", "CPU A then takes this interrupt and ..."

I'm having trouble understanding what the race is.  I believe you, I just don't see it.

Also: can you split the renaming (which is fine) into a separate patch to make the logic changes clear?

chen-png commented:
> k_timer timeout handler (`z_timer_experation_handler()`) is completely unsynchronized! 

@andyross yes, you are right, get the timer lock in this function to synchronize is the correct fix. I tested your change, it solved the issue.
and thanks for your help, I only know that it needs to use timer lock to sync, but I can't decide where it's the best place to put it, thanks your explaination.

Commit:
timer: mask interrupts in timer's timeout handler.

before running timer's timeout function, we need to make
sure that those threads waiting on this timer have been
added into the timer's wait queue, so add operations to
use timer lock to mask interrupts in z_timer_expiration_handler
function to synchronize timer's wait queue.

Signed-off-by: Chen Peng1 <peng1.chen@intel.com>

chen-png commented:
hi @andyross, could you help review this again? I updated the code according to your change, thanks.
